File: server/api/src/main/java/com/boxtrotstudio/ghost/common/network/packet/PingPongPacket.java
package com.boxtrotstudio.ghost.common.network.packet;

import com.boxtrotstudio.ghost.common.network.BasePlayerClient;
import com.boxtrotstudio.ghost.common.network.BaseServer;
import com.boxtrotstudio.ghost.network.packet.Packet;

import java.io.IOException;

public class PingPongPacket extends Packet<BaseServer, BasePlayerClient> {
    public PingPongPacket() {
        super();
    }
    public PingPongPacket(BasePlayerClient client) {
        super(client);
    }

    @Override
    protected void onHandlePacket(BasePlayerClient client) throws IOException {
        long ping = consume(8).asLong();

        client.onPing();
        new PingPongPacket(client).writePacket(ping);
    }

    @Override
    protected void onWritePacket(BasePlayerClient client, Object... args) throws IOException {
        long toSend = (long) args[0];

        client.getServer().sendUdpPacket(
            write((byte)0x09).
            write(toSend).
            write(new byte[24]).
            endUDP()
        );
    }
}

File: client/core/src/com/boxtrotstudio/ghost/client/network/packets/PingPongPacket.kt
package com.boxtrotstudio.ghost.client.network.packets

import com.boxtrotstudio.ghost.client.Ghost
import com.boxtrotstudio.ghost.client.network.Packet
import com.boxtrotstudio.ghost.client.network.PlayerClient

class PingPongPacket : Packet<PlayerClient>() {

    override fun write(vararg args: Any?) {
        val toSend = System.currentTimeMillis()

        write(0x09.toByte())
        write(toSend)
        write(ByteArray(24))
        endUdp()
    }

    override fun handle() {
        val pingStartTime = consume(8).asLong()

        val timeTook = System.currentTimeMillis() - pingStartTime
        Ghost.latency = timeTook
    }
}